Add tutorial for superdense coding (#1104)
vprusso/toqito: docs/tutorials.superdense_coding.rst

Superdense Coding
==================

In classical communication, sending two bits of information requires transmitting
two physical bits. But with the help of quantum mechanics we can bend this rule
using the phenomenon of quantum entanglement. **Superdense coding** proposed by
Bennet and Wiesner in 1992 :cite:`Bennett_1992_Communication` lets Alice send two classical bits to

> Bob by transmitting just *one qubit* . The catch here is that they must share an

Reviewer: ```suggestion
Bob by transmitting just *one qubit*. The catch here is that they must share an
```

entangled pair of qubits beforehand. We will explain this protocol in detail 
below:

> Superdense Coding protocol

Reviewer: ```suggestion
Superdense coding protocol
```

(to maintain consistency with the capitalization of things as you write them above).

^^^^^^^^^^^^^^^^^^^^^^^^^^
1. Before any communication begins, a third party prepares two qubits in 
   *Bell state*:

   .. math::

      \begin{equation}
          \begin{aligned}
              \ket{\psi} = \frac{\ket{00} + \ket{11}}{\sqrt{2}}
          \end{aligned}
      \end{equation}

   Alice takes the first qubit, Bob takes the second, and they both separate.
   This entangled pair is responsible for linking qubit *non-locally*, allowing
   Alice's local operations to affect the global state.

   .. code-block:: python

       import numpy as np
       from toqito.states import bell
       from toqito.matrices import pauli, cnot, hadamard
       
       bell_state = bell(0)
       print("Initial Bell state (|Φ⁺⟩):\n", bell_state)

> 2. Alice holds two classical bits (:math:`a` and :math:`b`) she wants to send.

Reviewer: ```suggestion
2. Alice holds two classical bits (:math:`a` and :math:`b`) that she wants to send.
```

   Depending on their values she applies one of four *Pauli Gates* to her qubit:

.. raw:: html

   <div style="text-align: center;">

.. list-table:: 
   :header-rows: 1
   :widths: 20 20 40 60 100
   
   * - :math:`a`
     - :math:`b`

>      - message

Reviewer: ```suggestion
     - *message*
```

>      - *Gate Applied*

Reviewer: ```suggestion
     - *Gate applied*
```

>      - *Final Output (Bell State)*

Reviewer: ```suggestion
     - *Final output (Bell state)*
```

   * - :math:`0`
     - :math:`0`
     - :math:`\ket{00}`
     - :math:`I`
     - :math:`\frac{|00\rangle + |11\rangle}{\sqrt{2}}`
   * - :math:`0`
     - :math:`1`
     - :math:`\ket{01}`
     - :math:`X`
     - :math:`\frac{|10\rangle + |01\rangle}{\sqrt{2}}`
   * - :math:`1`
     - :math:`0`
     - :math:`\ket{10}`
     - :math:`Z`
     - :math:`\frac{|00\rangle - |11\rangle}{\sqrt{2}}`
   * - :math:`1`
     - :math:`1`
     - :math:`\ket{11}`
     - :math:`XZ = iY`
     - :math:`\frac{|10\rangle - |01\rangle}{\sqrt{2}}`

.. raw:: html

   </div>

.. code-block:: python

>     pauli_gate_operations = {
>         "00": pauli("I"),  # Identity gate
>         "01": pauli("X"),  # Pauli-X gate
>         "10": pauli("Z"),  # Pauli-Z gate
>         "11": pauli("X") @ pauli("Z")  # X followed by Z (equivalent to iY)
>     }

Reviewer: [nit]:
```suggestion
    pauli_gate_operations = {
        # Identity gate.
        "00": pauli("I"),
        # Pauli-X gate.
        "01": pauli("X"),
        # Pauli-Z gate        
        "10": pauli("Z"),
        # X followed by Z (equivalent to iY)        
        "11": pauli("X") @ pauli("Z")
    }
```

    message_to_encode = "11"

    entangled_state_encoded = np.kron(pauli_gate_operations[message_to_encode], pauli("I")) @ bell_state

3. Bob performs operations to reverse the entanglement and extract the bits. First, 
   he applies a Controlled-NOT or :math:`CX` *(CNOT) Gate* with qubit received from Alice as the
   *control qubit* and Bob's original qubit as *target qubit*. After this, Bob moves
   ahead and applies a Hadamard or :math:`H` gate to Alice's qubit.

   .. code-block:: python

       state_after_cnot = cnot() @ entangled_state_encoded
       decoded_state = np.kron(hadamard(1), pauli("I")) @ state_after_cnot
       print("Decoded state:\n", decoded_state)

4. Finally, Bob measures both qubits in the computational basis (:math:`\ket{0}, 

>    \ket{1}`). The result is guaranteed to be :math:`ab`, the two bits Alice sent.

Reviewer: ```suggestion
   \ket{1}`). The result is guaranteed to be :math:`ab`, the two bits that Alice sent.
```

   .. code-block:: python

       measurement_probabilities = np.abs(decoded_state.flatten())**2
       print("Measurement probabilities for basis states |00>, |01>, |10>, |11>:")
       print(measurement_probabilities)

Full code :

.. code-block:: python

    import numpy as np
    from toqito.states import bell
    from toqito.matrices import pauli, cnot, hadamard

    bell_state = bell(0)
    print("Initial Bell state (|Φ⁺⟩):\n", bell_state)

> 

Reviewer: ```suggestion
```

    pauli_gate_operations = {
        "00": pauli("I"),
        "01": pauli("X"),
        "10": pauli("Z"),
        "11": pauli("X") @ pauli("Z")
    }

> 

Reviewer: ```suggestion
```

    message_to_encode = "11"

> 

Reviewer: ```suggestion
```

       from toqito.states import bell
       from toqito.matrices import pauli, cnot, hadamard
       
       bell_state = bell(0)
       print("Initial Bell state (|Φ⁺⟩):\n", bell_state)

2. Alice holds two classical bits (:math:`a` and :math:`b`) that she wants to send.
   Depending on their values she applies one of four *Pauli Gates* to her qubit:

.. raw:: html

   <div style="text-align: center;">

.. list-table:: 
   :header-rows: 1
   :widths: 20 20 40 60 100
   
   * - :math:`a`
     - :math:`b`
     - *message*
     - *Gate applied*
     - *Final output (Bell state)*
   * - :math:`0`
     - :math:`0`
     - :math:`\ket{00}`
     - :math:`I`
     - :math:`\frac{|00\rangle + |11\rangle}{\sqrt{2}}`
   * - :math:`0`
     - :math:`1`
     - :math:`\ket{01}`
     - :math:`X`
     - :math:`\frac{|10\rangle + |01\rangle}{\sqrt{2}}`
   * - :math:`1`
     - :math:`0`
     - :math:`\ket{10}`
     - :math:`Z`
     - :math:`\frac{|00\rangle - |11\rangle}{\sqrt{2}}`
   * - :math:`1`
     - :math:`1`
     - :math:`\ket{11}`
     - :math:`XZ = iY`
     - :math:`\frac{|10\rangle - |01\rangle}{\sqrt{2}}`

.. raw:: html

   </div>

.. code-block:: python

    pauli_gate_operations = {

>         # Identity gate.
> 
>         "00": pauli("I"),
> 
>         # Pauli-X gate.
> 
>         "01": pauli("X"),
> 
>         # Pauli-Z gate.
> 
>         "10": pauli("Z"),
> 
>         # X followed by Z (equivalent to iY).
> 
>         "11": pauli("X") @ pauli("Z")  
>     }

Reviewer: ```suggestion
        # Identity gate.
        "00": pauli("I"),
        # Pauli-X gate.
        "01": pauli("X"),
        # Pauli-Z gate.
        "10": pauli("Z"),
        # X followed by Z (equivalent to iY).
        "11": pauli("X") @ pauli("Z")  
    }
```

    message_to_encode = "11"

    entangled_state_encoded = np.kron(pauli_gate_operations[message_to_encode], pauli("I")) @ bell_state

3. Bob performs operations to reverse the entanglement and extract the bits. First, 

>    he applies a Controlled-NOT or :math:`CX` *(CNOT) Gate* with qubit received from Alice as the

Reviewer: ```suggestion
   he applies a Controlled-NOT or :math:`CX` *(CNOT) Gate* with the qubit received from Alice as the
```

>    *control qubit* and Bob's original qubit as *target qubit*. After this, Bob moves

Reviewer: ```suggestion
   *control qubit* and Bob's original qubit as the *target qubit*. After this, Bob moves
```

   ahead and applies a Hadamard or :math:`H` gate to Alice's qubit.

   .. code-block:: python

       state_after_cnot = cnot() @ entangled_state_encoded
       decoded_state = np.kron(hadamard(1), pauli("I")) @ state_after_cnot
       print("Decoded state:\n", decoded_state)

4. Finally, Bob measures both qubits in the computational basis (:math:`\ket{0}, 

>    \ket{1}`). The result is guaranteed to be :math:`ab`, the two bits that Alice sent.

Reviewer: ```suggestion
   \ket{1}`). The result is guaranteed to be :math:`ab`; the two bits that Alice sent.
```

   .. code-block:: python

       measurement_probabilities = np.abs(decoded_state.flatten())**2
       print("Measurement probabilities for basis states |00>, |01>, |10>, |11>:")
       print(measurement_probabilities)

> Full code :

Reviewer: ```suggestion
Full code:
```

Superdense Coding
==================

In classical communication, sending two bits of information requires transmitting

> two physical bits. But with the help of quantum mechanics we can bend this rule
> using the phenomenon of quantum entanglement. **Superdense coding** proposed by

Reviewer 2: ```suggestion
two physical bits. But with the help of quantum entanglement, we can bend this rule. **Superdense coding** proposed by
```

Reviewer 2: move the line `**Superdense coding** proposed by` and everything after this to a new paragraph.

Bennet and Wiesner in 1992 :cite:`Bennett_1992_Communication` lets Alice send two classical bits to
Bob by transmitting just *one qubit*. The catch here is that they must share an
entangled pair of qubits beforehand. We will explain this protocol in detail 
below:

Superdense coding protocol
^^^^^^^^^^^^^^^^^^^^^^^^^^
1. Before any communication begins, a third party prepares two qubits in 
   *Bell state*:

   .. math::

      \begin{equation}
          \begin{aligned}
              \ket{\psi} = \frac{\ket{00} + \ket{11}}{\sqrt{2}}
          \end{aligned}
      \end{equation}

   Alice takes the first qubit, Bob takes the second, and they both separate.
   This entangled pair is responsible for linking qubit *non-locally*, allowing
   Alice's local operations to affect the global state.

   .. code-block:: python

       import numpy as np
       from toqito.states import bell
       from toqito.matrices import pauli, cnot, hadamard
       
       bell_state = bell(0)
       print("Initial Bell state (|Φ⁺⟩):\n", bell_state)

2. Alice holds two classical bits (:math:`a` and :math:`b`) that she wants to send.

>    Depending on their values she applies one of four *Pauli Gates* to her qubit:

Reviewer 2: ```suggestion
   Depending on the values of her classical bits, she applies one of the four *Pauli Gates* to her qubit for encoding:
```

.. raw:: html

   <div style="text-align: center;">

.. list-table:: 
   :header-rows: 1
   :widths: 20 20 40 60 100
   
   * - :math:`a`
     - :math:`b`
     - *message*
     - *Gate applied*
     - *Final output (Bell state)*
   * - :math:`0`
     - :math:`0`
     - :math:`\ket{00}`
     - :math:`I`
     - :math:`\frac{|00\rangle + |11\rangle}{\sqrt{2}}`
   * - :math:`0`
     - :math:`1`
     - :math:`\ket{01}`
     - :math:`X`
     - :math:`\frac{|10\rangle + |01\rangle}{\sqrt{2}}`
   * - :math:`1`
     - :math:`0`
     - :math:`\ket{10}`
     - :math:`Z`
     - :math:`\frac{|00\rangle - |11\rangle}{\sqrt{2}}`
   * - :math:`1`
     - :math:`1`
     - :math:`\ket{11}`
     - :math:`XZ = iY`
     - :math:`\frac{|10\rangle - |01\rangle}{\sqrt{2}}`

.. raw:: html

>    </div>

Reviewer 2: What are these code blocks doing?

Author: These code blocks are rendering tables inside the `.rst` file

Reviewer 2: I don't think you need those. 

https://sublime-and-sphinx-guide.readthedocs.io/en/latest/tables.html#list-table-directive

Author: > I don't think you need these.
> 
> https://sublime-and-sphinx-guide.readthedocs.io/en/latest/tables.html#list-table-directive

I tried this only first, but this leads to table not getting centered correctly, for which I used html blocks and hence the line:
```rst
.. raw:: html

   <div style="text-align: center;">

.. list-table:: 
```
at start

Reviewer 2: Doesn't this work?

```
.. list-table::
   :align: center
```

.. code-block:: python

    pauli_gate_operations = {
        # Identity gate.
        "00": pauli("I"),
        # Pauli-X gate.
        "01": pauli("X"),
        # Pauli-Z gate.
        "10": pauli("Z"),
        # X followed by Z (equivalent to iY).

>         "11": pauli("X") @ pauli("Z")  

Reviewer 2: technically speaking you can use complex i times pauli Y here.

    }

    message_to_encode = "11"

>     entangled_state_encoded = np.kron(pauli_gate_operations[message_to_encode], pauli("I")) @ bell_state

Reviewer 2: Missing the step where Alice sends her encoded message.

Author: ```python

    message_to_encode = "11"
    
    #Alice sends her encoded entangeled state after this step.
    entangled_state_encoded = np.kron(pauli_gate_operations[message_to_encode], pauli("I")) @ bell_state
```

I have added this comment

> 3. Bob performs operations to reverse the entanglement and extract the bits. First, 

Reviewer 2: reverse the entanglement on what? See my previous suggestion.

Author: I have added 
```text
Bob performs operations to reverse the entanglement on encoded state sent by Alice and extract the bits.
```

   he applies a Controlled-NOT or :math:`CX` *(CNOT) Gate* with the qubit received from Alice as the
   *control qubit* and Bob's original qubit as the *target qubit*. After this, Bob moves
   ahead and applies a Hadamard or :math:`H` gate to Alice's qubit.

   .. code-block:: python

       state_after_cnot = cnot() @ entangled_state_encoded
       decoded_state = np.kron(hadamard(1), pauli("I")) @ state_after_cnot
       print("Decoded state:\n", decoded_state)

4. Finally, Bob measures both qubits in the computational basis (:math:`\ket{0}, 

>    \ket{1}`). The result is guaranteed to be :math:`ab`; the two bits that Alice sent.

Reviewer 2: ```suggestion
   \ket{1}`). The result is guaranteed to be :math:`11`; the two bits that Alice sent.
```

       print("Measurement probabilities for basis states |00>, |01>, |10>, |11>:")
       print(measurement_probabilities)

Full code:

.. code-block:: python

  >>> import numpy as np
  >>> from toqito.states import bell
  >>> from toqito.matrices import pauli, cnot, hadamard
  >>>
  >>> bell_state = bell(0)
  >>> print("Initial Bell state (|Φ⁺⟩):\n", bell_state)
  Initial Bell state (|Φ⁺⟩):
   [[0.70710678]
   [0.        ]
   [0.        ]
   [0.70710678]]

  >>> pauli_gate_operations = {
  ...     "00": pauli("I"),
  ...     "01": pauli("X"),
  ...     "10": pauli("Z"),
  ...     "11": pauli("X") @ pauli("Z")
  ... }

  >>> message_to_encode = "10"

  >>> entangled_state_encoded = np.kron(pauli_gate_operations[message_to_encode], pauli("I")) @ bell_state

  >>> state_after_cnot = cnot() @ entangled_state_encoded

  >>> decoded_state = np.kron(hadamard(1), pauli("I")) @ state_after_cnot

  >>> measurement_probabilities = np.abs(decoded_state.flatten())**2
  >>> print("Measurement probabilities for basis states |00>, |01>, |10>, |11>:\n", measurement_probabilities)
  Measurement probabilities for basis states |00>, |01>, |10>, |11>:

>    [0. 0. 1. 0.]

Reviewer 2: Do we need this full code block?

Superdense Coding
==================

In classical communication, sending two bits of information requires transmitting
two physical bits. But with the help of quantum entanglement, we can bend this rule. 

> **Superdense coding** proposed bye coding** proposed by Bennet and Wiesner in 1992 

Reviewer 2: ```suggestion
**Superdense coding** proposed by Bennet and Wiesner in 1992 
```
typo

:cite:`Bennett_1992_Communication` lets Alice send two classical bits to
Bob by transmitting just *one qubit*. The catch here is that they must share an
entangled pair of qubits beforehand. We will explain this protocol in detail 
below:

Superdense coding protocol
^^^^^^^^^^^^^^^^^^^^^^^^^^
1. Before any communication begins, a third party prepares two qubits in 
   *Bell state*:

   .. math::

      \begin{equation}
          \begin{aligned}
              \ket{\psi} = \frac{\ket{00} + \ket{11}}{\sqrt{2}}
          \end{aligned}
      \end{equation}

   Alice takes the first qubit, Bob takes the second, and they both separate.

>    This entangled pair is responsible for linking qubit *non-locally*, allowing

Reviewer 2: ```suggestion
   This entangled pair is responsible for linking the qubits *non-locally*, allowing
```

   Alice's local operations to affect the global state.

   .. code-block:: python

       import numpy as np
       from toqito.states import bell
       from toqito.matrices import pauli, cnot, hadamard
       
       bell_state = bell(0)
       print("Initial Bell state (|Φ⁺⟩):\n", bell_state)

> 2. Alice holds two classical bits (:math:`a` and :math:`b`) that she wants to send.

Reviewer 2: ```suggestion
2. Alice holds two classical bits (:math:`a` and :math:`b`) that she wants to send. For the tutorial, she is choosing to send :math:`11`.
```

   to her qubit for encoding:

.. raw:: html

   <div style="text-align: center;">

.. list-table:: 
   :header-rows: 1
   :widths: 20 20 40 60 100
   
   * - :math:`a`
     - :math:`b`
     - *message*
     - *Gate applied*
     - *Final output (Bell state)*
   * - :math:`0`
     - :math:`0`
     - :math:`\ket{00}`
     - :math:`I`
     - :math:`\frac{|00\rangle + |11\rangle}{\sqrt{2}}`
   * - :math:`0`
     - :math:`1`
     - :math:`\ket{01}`
     - :math:`X`
     - :math:`\frac{|10\rangle + |01\rangle}{\sqrt{2}}`
   * - :math:`1`
     - :math:`0`
     - :math:`\ket{10}`
     - :math:`Z`
     - :math:`\frac{|00\rangle - |11\rangle}{\sqrt{2}}`
   * - :math:`1`
     - :math:`1`
     - :math:`\ket{11}`
     - :math:`XZ = iY`
     - :math:`\frac{|10\rangle - |01\rangle}{\sqrt{2}}`

.. raw:: html

   </div>

.. code-block:: python

    pauli_gate_operations = {
        # Identity gate.
        "00": pauli("I"),
        # Pauli-X gate.
        "01": pauli("X"),
        # Pauli-Z gate.
        "10": pauli("Z"),
        # X followed by Z (equivalent to iY).
        "11": 1j * pauli("Y")  
    }

    message_to_encode = "11"
    

>     #Alice sends her encoded entangeled state after this step.

Reviewer 2: ```suggestion
    # Alice sends her encoded entangled state after this step.
```

   .. code-block:: python

>        >>> state_after_cnot = cnot() @ entangled_state_encoded

Reviewer 2: @RajdeepAher Why did you change the code blocks?

The code blocks are not being executed with the version after this commit. https://toqito--1104.org.readthedocs.build/en/1104/tutorials.superdense_coding.html

Author: Hello, I made changes because I wanted the code to be divided into chunks that correspond to each step.

However, when I add them with the >>> indicator and include print statements for verification, it triggers doctests, causing errors due to undefined variable names in each chunk. To resolve this, I attempted to organize the chunks recursively, but that made the code look messier.

If I remove the print statements, it ends up being the same as what I did in the latest commit, but without the >>>.

I tried to fix it, but I’m not sure how to proceed.

Reviewer 2: IDK what the issue is. I haven't added a tutorial for Toqito yet. 

If I look at how the other tutorials are formatted, it looks like @vprusso hard-coded the output results. Maybe we need to find a way to execute the code blocks and add the results to the tutorials?

Author: I can think of a solution in which I print outputs for the last part (full code) and verify it with doctests. Would that work?

Reviewer: Agreed that not hardcoding the results are the way to go, and I agree with the issue that @purva-thakre wrote up in #1108. Out of scope for this issue, but I think that we should be dynamically outputting the results of the code in the examples for sure.

Reviewer: > I can think of a solution in which I print outputs for the last part (full code) and verify it with doctests. Would that work?

I think this makes sense for this notebook and MR. We can save the more thorough solution for the issue that @purva-thakre wrote up in #1108

Author: Hello @vprusso 

> > I can think of a solution in which I print outputs for the last part (full code) and verify it with doctests. Would that work?
> 
> I think this makes sense for this notebook and MR. We can save the more thorough solution for the issue that @purva-thakre wrote up in #1108

I have implemented it in https://github.com/vprusso/toqito/pull/1104/commits/17cc68d3776065b1dd8a5d30d937d242939e550e , please check.
